smart-contracts/contracts/utils/UnlockSwapBurner.sol

```diff
@@ -0,0 +1,148 @@
+   * @param _permit2Address the address of Uniswap PERMIT2 contract
+   */
+  constructor(
+    address _unlockAddress,
+    address _permit2Address,
+    address _uniswapRouter
+  ) {
+    unlockAddress = _unlockAddress;
+    permit2 = _permit2Address;
+    uniswapRouter = _uniswapRouter;
+  }
+
+  /**
+   * Simple helper to retrieve balance in ERC20 or native tokens
+   * @param token the address of the token (address(0) for native token)
+   */
+  function getBalance(address token) internal view returns (uint) {
+    return
+      token == address(0)
+        ? address(this).balance
+        : IMintableERC20(token).balanceOf(address(this));
+  }
+
+  /**
+   * Swap tokens to UDT and burn the tokens
+   */
+  function swapAndBurn(
+    address tokenAddress,
+    uint24 poolFee
+  ) public payable returns (uint amount) {
+    // get info from unlock
+    address udtAddress = IUnlock(unlockAddress).udt();
+    address wrappedAddress = IUnlock(unlockAddress).weth();
+
+    // get total balance of token to swap
+    uint tokenAmount = getBalance(tokenAddress);
+
+    if (tokenAddress == udtAddress) {
+      revert UnauthorizedSwap();
+    }
+
+    // wrap native tokens
+    if (tokenAddress == address(0)) {
+      IWETH(wrappedAddress).deposit{value: tokenAmount}();
+      tokenAddress = wrappedAddress;
+      tokenAmount = getBalance(tokenAddress);
+    }
+
+    // approve ERC20 spending
+    if (tokenAddress != address(0)) {
+      // Approve the router to spend src ERC20
+      TransferHelper.safeApprove(tokenAddress, uniswapRouter, tokenAmount);
+
+      // approve PERMIT2 to manipulate the token
+      IERC20(tokenAddress).approve(permit2, tokenAmount);
+    }
+
+    // issue PERMIT2 Allowance
+    IPermit2(permit2).approve(
+      tokenAddress,
+      uniswapRouter,
+      tokenAmount.toUint160(),
+      uint48(block.timestamp + 60) // expires after 1min
+    );
+
+    bytes memory defaultPath = abi.encodePacked(
+      wrappedAddress,
+      uint24(3000), // default UDT pool fee is set to 0.3%
+      udtAddress
+    );
+
+    // executes the swap token > WETH > UDT
+    ISwapRouter.ExactInputParams memory params = ISwapRouter.ExactInputParams({
+      path: tokenAddress == wrappedAddress
+        ? defaultPath
+        : abi.encodePacked(tokenAddress, poolFee, defaultPath),
+      recipient: address(this),
+      deadline: block.timestamp + 60, // expires after 1min
+      amountIn: tokenAmount,
+      amountOutMinimum: 0
+    });
+
+    // Executes the swap.
+    uint amountUDTOut = ISwapRouter(uniswapRouter).exactInput(params);
+    if (amountUDTOut == 0) {
+      revert UDTSwapFailed(uniswapRouter, tokenAddress, tokenAmount);
+    }
+
+    // burn the UDT
+    bool success = IERC20(udtAddress).transfer(burnAddress, amountUDTOut);
+    if (success == false) {
+      revert UDTSwapFailed(uniswapRouter, tokenAddress, tokenAmount);
+    } else {
+      emit SwapBurn(tokenAddress, tokenAmount, amountUDTOut);
```

Comment:
Can we emit the same event as the eRC20 contract would maybe?

Reply:
That will be a `Transfer` event. Just will make it harder to read maybe (as the erc20 contract will emit one anyway)?

Reply:
I was wondering if there was a specific `Burn` event in ERC20 contracts, but maybe not... in an y case, nbd!

Reply:
No its usually a simple transfer to address zero (here to "dead" address)